notifications/hub.go

```diff
@@ -0,0 +1,61 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package notifications
+
+import (
+	"sync"
+)
+
+// A Hub allows subscribers to receive notifications.
+type Hub struct {
+	sync.Mutex
+	subscribers Subscribers
+}
+
+// Subscribe registers a subscriber to receive notifications.
+func (h *Hub) Subscribe(s *Subscriber) {
+	h.Lock()
+	defer h.Unlock()
+
+	h.subscribers = append(h.subscribers, s)
+}
+
+// Unsubscribe unregisters a subscriber from notifications.
+func (h *Hub) Unsubscribe(s *Subscriber) {
+	h.Lock()
+	defer h.Unlock()
+
+	for i, sub := range h.subscribers {
+		if s == sub {
+			h.subscribers = append(h.subscribers[:i], h.subscribers[i+1:]...)
+			break
```

Comment:
You're correct that a subscriber could be subscribed twice actually, good spot! I added the `uuid` field to `Subscriber` in order to tell one from another when removing but never considered the double subscribe scenario, I'll amend this.